Binary search based IpSubnetFilter

hyperxpro opened the pull request:
Motivation:

`IpSubnetFilter` uses Binary Search for IP Address search which is fast if we have large set of IP addresses to filter.

Modification:

Added `IpSubnetFilter` which takes `IpSubnetFilterRule` for filtering.

Result:
Faster IP address filter.

hyperxpro commented:
For 50k IP Addresses, here is the benchmark:

The Target IP address was in the middle.

```
Benchmark                    Mode  Cnt         Score         Error  Units
BenchmarkPerf.binarySearch  thrpt    5  12235564.910 ± 4653778.680  ops/s
BenchmarkPerf.forLoop       thrpt    5      4494.061 ±    1570.423  ops/s
```

normanmaurer commented on handler/src/main/java/io/netty/handler/ipfilter/IpSubnetFilter.java:
```diff
@@ -0,0 +1,74 @@
+
+    @Override
+    protected boolean accept(ChannelHandlerContext ctx, InetSocketAddress remoteAddress) {
+        int indexOf = Collections.binarySearch(this.rules, remoteAddress, new IpSubnetFilterRuleComparator());
```
We should consider adding `IpSubnetFilterRuleComparator.INSTANCE`, mark the constructor private and reuse the same instance everywhere

hyperxpro replied:
I'll do that.

chrisvest commented on handler/src/main/java/io/netty/handler/ipfilter/IpSubnetFilterRuleComparator.java:
```diff
@@ -0,0 +1,34 @@
+                return ((IpSubnetFilterRule) o1).matches((InetSocketAddress) o2) ? 0 : -1;
+            }
+        }
+        return -1;
```
This is a rather strange comparator, that skimps on the reciprocity rule, and only kind of makes sense in this narrow use case. And then it also relies on the implementation details of `Collections.binarySearch` a bit. It's also not clear if it actually works as intended. I'd rather see a bespoke and well tested implementation of binary search as a helper method for `IpSubnetFilter.accept`, that incorporates the special comparison logic.

hyperxpro replied:
Binary search requires me to pass same class type as key. So if I wanna search InetSocketAddress in list of IpSubnetFilterRule, it's not possible. So I used Object here.

o1 will always be IpSubnetFilterRule.
o2 will be:

1. IpSubnetFilterRule if we're sorting. (It'll use networkAddress for sort)

2. InetSocketAddress if we're searching. (It'll simple use "matches' method). 

You have a better approach for this?

chrisvest replied:
Yes, this particular use of binary search is a little unusual. That's why I suggested to either reimplement a bespoke binary search, that captures and handles the special comparison logic. Or, since the `Collections.sort` method has an overload that relies on `Comparable` rather than `Comparator`, we move or inline this `Comparator` to the `accept` (the only place it's used at that point) and simplify it (since o1 is now always a filter, and o2 is always an address).

chrisvest replied:
Also, I'd like to see a unit test that verifies that the `accept` method filters correctly; so everything works as intended.

franz1981 commented:
Too much variance I am not convinced the bench is measuring things correctly!

chrisvest commented on handler/src/main/java/io/netty/handler/ipfilter/IpSubnetFilterRule.java:
```diff
@@ -68,20 +68,35 @@ public IpFilterRuleType ruleType() {
+
     private static final class Ip4SubnetFilterRule implements IpFilterRule {
 
+        private final BigInteger networkAddressAsBigInteger;
```
I wonder if we can implement a comparison without needing to keep this around, or if that adds too much overhead.

hyperxpro replied:
Yeah, I'll use Integer directly here now.

chrisvest reviewed (commented):
Had some comments.

hyperxpro commented:
> Too much variance I am not convinced the bench is measuring things correctly!

Even I thought the same when I looked at it for the first time. To double check, I added try catch block to catch ArrayIndexOutOfFound exception to see if my binary search was failing. However, no exceptions at all.

And for-loop for 25k (my target entry was in middle of 50k list) seems to be slow enough.

hyperxpro commented on handler/src/main/java/io/netty/handler/ipfilter/IpSubnetFilter.java:
```diff
@@ -0,0 +1,74 @@
+            this.rules.add(ipSubnetFilterRule);
+        }
+
+        Collections.sort(this.rules, new IpSubnetFilterRuleComparator());
```
Collections#binarySearch requires Comparator. Can't skip.

chrisvest replied:
I mean, for the `sort` method calls.

franz1981 commented:
Try measuring branch mispredict with perf norm counters or by inspecting GC profiling...there must be a reason for the variance

hyperxpro commented:
Ran some random IP test and now Binary search is showing extreme results.

I inserted 1.1.1.1 in the list, sorted along with other random IP addresses (generated using java.util.Random).

Binary search is skipping long indexes.

```
16843009 against -392714775
Failed for: /1.1.1.1:9110 > IpSubnetFilterRule{ipAddress='232.151.165.233'}
16843009 against 712155077
Failed for: /1.1.1.1:9110 > IpSubnetFilterRule{ipAddress='42.114.159.197'}
```

My 1.1.1.1 address was in the middle of these 2 entries (along with other IP addresses like 6.X.X.X, 13.X.X.X, etc).

Not sure about this.

chrisvest commented on handler/src/main/java/io/netty/handler/ipfilter/IpSubnetFilter.java:
```diff
@@ -0,0 +1,150 @@
+                // If parentRule is null, take first element out of Iterator.
+                if (parentRule == null) {
+                    parentRule = iterator.next();
+                }
```
Please do another `hasNext` check, instead of relying on catching `NoSuchElementException`.

hyperxpro replied:
`hasNext` is already done in `while(Iterator#hasNext())` so no need to do the same here but I'll do it in other ones.

chrisvest commented on handler/src/main/java/io/netty/handler/ipfilter/IpSubnetFilter.java:
```diff
@@ -0,0 +1,150 @@
+
+                // If parentRule matches childRule, schedule that Rule for deletion.
+                if (parentRule.matches(new InetSocketAddress(childRule.getIpAddress(), 1))) {
+                    toRemove.add(childRule);
```
Instead of collecting a list of things to remove (or removing via `Iterator.remove()`), you can instead build up a new list of things you want to keep. This algorithm would be linear in time, instead of N*M from `removeAll`. You can then do the null-checking at the same time. Make this method take the list-to-be-filtered as a parameter, and return the new filtered list. Then you can do all of this at the `this.rules` field assignment.

hyperxpro replied:
I added code that checks if all rules belong to the same type while checking for `null` rules inside the user-supplied list of rules. If I try to move these inside sorting codes, it'll be a complete mess. 

The reason behind adding this: We don't have to call `List#get(int)` because we already know if the list of rule contains the same `IpFilterRuleType`.

PTAL at the latest code and revert back. :D

chrisvest replied:
I mean something like this:

```patch
Index: handler/src/main/java/io/netty/handler/ipfilter/IpSubnetFilter.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- handler/src/main/java/io/netty/handler/ipfilter/IpSubnetFilter.java	(revision 295e104c7ed9027584927777073ae9f2f81cc2df)
+++ handler/src/main/java/io/netty/handler/ipfilter/IpSubnetFilter.java	(date 1598435434667)
@@ -86,14 +86,14 @@
      * @param rules            {@link IpSubnetFilterRule} as {@link List}
      */
     public IpSubnetFilter(boolean acceptIfNotFound, List<IpSubnetFilterRule> rules) {
-        this.rules = ObjectUtil.checkNotNull(rules, "rules");
+        ObjectUtil.checkNotNull(rules, "rules");
         this.acceptIfNotFound = acceptIfNotFound;
 
         int numAccept = 0;
         int numReject = 0;
 
         // Iterate over rules and check for `null` rule.
-        for (IpSubnetFilterRule ipSubnetFilterRule : this.rules) {
+        for (IpSubnetFilterRule ipSubnetFilterRule : rules) {
             ObjectUtil.checkNotNull(ipSubnetFilterRule, "rule");
             if (ipSubnetFilterRule.ruleType() == IpFilterRuleType.ACCEPT) {
                 numAccept++;
@@ -120,7 +120,7 @@
             ipFilterRuleType = null;
         }
 
-        sortAndFilter();
+        this.rules = sortAndFilter(rules);
     }
 
     @Override
@@ -143,39 +143,27 @@
      *     <li> Sort the list again </li>
      * </ol>
      */
-    private void sortAndFilter() {
-        Collections.sort(this.rules);
+    private static List<IpSubnetFilterRule> sortAndFilter(List<IpSubnetFilterRule> rules) {
+        Collections.sort(rules);
         Iterator<IpSubnetFilterRule> iterator = rules.iterator();
-        List<IpSubnetFilterRule> toRemove = new ArrayList<IpSubnetFilterRule>();
+        List<IpSubnetFilterRule> toKeep = new ArrayList<IpSubnetFilterRule>();
 
-        IpSubnetFilterRule parentRule = null;
+        IpSubnetFilterRule parentRule = iterator.hasNext() ? iterator.next() : null;
+        if (parentRule != null) {
+            toKeep.add(parentRule);
+        }
         while (iterator.hasNext()) {
-
-            // If parentRule is null, take first element out of Iterator.
-            if (parentRule == null) {
-                parentRule = iterator.next();
-            }
-
-            // If we don't have any more rule, we're done now.
-            if (!iterator.hasNext()) {
-                break;
-            }
-
-            // Take one more element out of Iterator, this will be childRule.
+            // Grab a potential child rule.
             IpSubnetFilterRule childRule = iterator.next();
 
-            // If parentRule matches childRule, schedule that Rule for deletion.
-            if (parentRule.matches(new InetSocketAddress(childRule.getIpAddress(), 1))) {
-                toRemove.add(childRule);
-            } else {
-                // If parentRule does not matches childRule, this childRule will become new parentRule
-                // and we'll do the same again
+            // If parentRule matches childRule, then there's no need to keep the child rule.
+            // Otherwise, the rules are distinct and we need both.
+            if (!parentRule.matches(new InetSocketAddress(childRule.getIpAddress(), 1))) {
+                toKeep.add(childRule);
+                // Then we'll keep the child rule around as the parent for the next round.
                 parentRule = childRule;
             }
         }
-
-        rules.removeAll(toRemove);
-        toRemove.clear();
-        Collections.sort(rules); // Re-sort just to be sure
+        return toKeep;
     }
 }
```

chrisvest replied:
@hyperxpro Not sure if I have access to directly push to this PR, but I can give it a go if you like.

hyperxpro replied:
Just create a gist of updated source and I will copy from there. :p

chrisvest replied:
If you use Intellij you can also copy the snippet above and click VCS→Apply Patch from Clipboard…

hyperxpro replied:
NVM, Did it myself. Thanks for helping a noob. :p

chrisvest commented on handler/src/main/java/io/netty/handler/ipfilter/IpSubnetFilter.java:
```diff
@@ -0,0 +1,150 @@
+
+        rules.removeAll(toRemove);
+        toRemove.clear();
+        Collections.sort(rules); // Re-sort just to be sure
```
If you follow my comment above, this part would also go away.

chrisvest reviewed (commented):
Couple of comments on the `sortAndFilter`.

franz1981 commented on handler/src/main/java/io/netty/handler/ipfilter/IpSubnetFilterRule.java:
```diff
@@ -68,6 +71,43 @@ public IpFilterRuleType ruleType() {
+    @Override
+    public int compareTo(IpSubnetFilterRule ipSubnetFilterRule) {
+        if (filterRule instanceof Ip4SubnetFilterRule) {
+            return Integer.valueOf(((Ip4SubnetFilterRule) filterRule).networkAddress)
```
I see that networkAddress is already an `int`, why you are translating it into an `Integer` just to be unboxed as an `int` again?

hyperxpro replied:
Sorry, forgot to take a look here.

`Integer#compareTo(int, int)` is documented for 1.7+. I get this error. So I've to call `Integer#value(int)` to do it in the old style.

hyperxpro replied:
As per https://github.com/netty/netty#how-to-build, We need Java 7 or higher to build Netty. But why IntelliJ is set to Java 6?

chrisvest replied:
Yeah, Netty 4.1 is meant to work on Java 6 at runtime.

hyperxpro replied:
That's the thing. Can't use `Integer#compareTo(int, int)` in Java 6. Maybe I should manually implement `Integer#compareTo(int, int)`. How does it sound?

franz1981 replied:
+1

hyperxpro replied:
Done

hyperxpro commented:
```
Benchmark                                                     Mode  Cnt         Score         Error   Units
BenchmarkPerf.binarySearch                                   thrpt   25  26404462.537 ±  938167.338   ops/s
BenchmarkPerf.binarySearch:·gc.alloc.rate                    thrpt   25        ≈ 10⁻⁴                MB/sec
BenchmarkPerf.binarySearch:·gc.alloc.rate.norm               thrpt   25        ≈ 10⁻⁶                  B/op
BenchmarkPerf.binarySearch:·gc.count                         thrpt   25           ≈ 0                counts
BenchmarkPerf.binarySearchAndGetElement                      thrpt   25  22392144.426 ± 1068876.685   ops/s
BenchmarkPerf.binarySearchAndGetElement:·gc.alloc.rate       thrpt   25        ≈ 10⁻⁴                MB/sec
BenchmarkPerf.binarySearchAndGetElement:·gc.alloc.rate.norm  thrpt   25        ≈ 10⁻⁶                  B/op
BenchmarkPerf.binarySearchAndGetElement:·gc.count            thrpt   25           ≈ 0                counts
BenchmarkPerf.forLoop                                        thrpt   25    476466.294 ±   15702.361   ops/s
BenchmarkPerf.forLoop:·gc.alloc.rate                         thrpt   25        ≈ 10⁻⁴                MB/sec
BenchmarkPerf.forLoop:·gc.alloc.rate.norm                    thrpt   25        ≈ 10⁻⁴                  B/op
BenchmarkPerf.forLoop:·gc.count                              thrpt   25           ≈ 0                counts

```

@franz1981 Boooooooom :p

hyperxpro commented:
@chrisvest 

There is 1 more issue with this.

108.0.0.0/4 has ranges from 96.0.0.0-111.255.255.255. It someone inserted this rule:
```
101.0.0.0/8
105.0.0.0/8
108.0.0.0/4
109.0.0.0/8
110.0.0.0/8
```
Current code will only remove: 
```
108.0.0.0/4
109.0.0.0/8
110.0.0.0/8
```

Of course, I will blame the programmer for adding such over-lapping rules. But just to be sure from our side, should we add a backward CIDR overlapping check. The current code checks the forward direction.

Your thoughts?

chrisvest commented:
@hyperxpro I put your example in a test, and it passes though?

```java
    @Test
    public void testSortAndFilter() {
        List<IpSubnetFilterRule> rules = new ArrayList<IpSubnetFilterRule>();
        IpSubnetFilterRule superRule;
        rules.add(buildRejectIP("101.0.0.0", 8));
        rules.add(buildRejectIP("105.0.0.0", 8));
        rules.add(superRule = buildRejectIP("108.0.0.0", 4));
        rules.add(buildRejectIP("109.0.0.0", 8));
        rules.add(buildRejectIP("110.0.0.0", 8));
        rules = IpSubnetFilter.sortAndFilter(rules);
        assertEquals(Arrays.asList(superRule), rules);
    }
```

hyperxpro commented:
That's not the issue. It'll pass. But if you iterate the list, you'll find 105.0.0.0/8 IP however, it should be there because 108.0.0.0/4 already covers it.

chrisvest commented:
I don't understand. The test above asserts on the contents of the whole list, and after `sortAndFilter` it only has the `108.0.0.0/4` rule in it.

hyperxpro commented:
See, when we insert 108.0.0.0/4 and then we insert 109.0.0.0/8, sortAndFilter will remove 109.0.0.0/8 since it overlaps with 108.0.0.0/4. However, sortAndFilter don't remove 105.0.0.0/8 even though it overlaps with 108.0.0.0/4. sortAndFilter needs backward scanning too in order to remove 105.0.0.0/8.

chrisvest commented:
No, it _is_ removed because 108.0.0.0/4 (`0x60000000`) is ordered before 105.0.0.0/8 (`0x69000000`) by the sort. The rules are sorted by their _masked_ addresses. This is what the test shows.

hyperxpro commented:
My PC is updating (typical Windows stuff), once it finishes. I've some code that reproduces this issue. I'll drop the output in sometime.

hyperxpro commented:
NVM, The old test was using the wrong code. Things are perfect now. 

@normanmaurer Ready for your review.

chrisvest commented on handler/src/test/java/io/netty/handler/ipfilter/IpSubnetFilterTest.java:
```diff
@@ -141,6 +144,52 @@ public void testUniqueIpFilterHandler() {
+        Assert.assertFalse(ch6.isActive());
+        Assert.assertTrue(ch6.close().isSuccess());
+
+        ipSubnetFilterRuleList.clear(); // Because we're done with the list
```
We don't need to explicitly clear this list. The GC will take care of it.

chrisvest commented on handler/src/test/java/io/netty/handler/ipfilter/IpSubnetFilterTest.java:
```diff
@@ -141,6 +144,52 @@ public void testUniqueIpFilterHandler() {
+        ipSubnetFilterRuleList.add(buildRejectIP("1.1.1.1", 8));
+        ipSubnetFilterRuleList.add(buildRejectIP("200.200.200.200", 32));
+        ipSubnetFilterRuleList.add(buildRejectIP("108.0.0.0", 4));
+        ipSubnetFilterRuleList.add(buildRejectIP("10.10.10.10", 8));
```
Would be good to also test IPv6, including testing with rule lists that is a mix of IPv4 and IPv6, and have them match on both IPv4 and IPv6 addresses.

hyperxpro replied:
And here we go again, Comparator tries to compare IPv4 and IPv6. How about explicitly defining the IP version type?

chrisvest replied:
Yeah, that could work. If we ensure that the list never contain both types, it would side-step a lot of trouble.

chrisvest replied:
The comparator, or the accept method, would still need to test that the address being checked is of the correct type, though. But that should be simple enough.

hyperxpro replied:
Added IPv4/6 count check. It'll throw an exception like this:
```java
java.lang.IllegalArgumentException: Mixed IP version rules are not allowed. Number of IPv4 rules: 5, IPv6 rules: 1
```

hyperxpro replied:
@chrisvest PTAL at the latest code.

chrisvest commented on handler/src/main/java/io/netty/handler/ipfilter/IpSubnetFilter.java:
```diff
@@ -0,0 +1,174 @@
+        }
+
+        this.rules = sortAndFilter(rules);
+        rules.clear(); // Free-up old unsorted list
```
I don't think this clear() is necessary. GC should take care of it.

chrisvest commented on handler/src/main/java/io/netty/handler/ipfilter/IpSubnetFilter.java:
```diff
@@ -0,0 +1,174 @@
+
+    @Override
+    protected boolean accept(ChannelHandlerContext ctx, InetSocketAddress remoteAddress) {
+        int indexOf = Collections.binarySearch(this.rules, remoteAddress, IpSubnetFilterRuleComparator.INSTANCE);
```
`this` is unnecessary.

chrisvest commented on handler/src/main/java/io/netty/handler/ipfilter/IpSubnetFilter.java:
```diff
@@ -0,0 +1,174 @@
+        int indexOf = Collections.binarySearch(this.rules, remoteAddress, IpSubnetFilterRuleComparator.INSTANCE);
+        if (indexOf >= 0) {
+            if (ipFilterRuleType == null) {
+                return this.rules.get(indexOf).ruleType() == IpFilterRuleType.ACCEPT;
```
`this` is unnecessary.

chrisvest commented on handler/src/main/java/io/netty/handler/ipfilter/IpSubnetFilter.java:
```diff
@@ -0,0 +1,174 @@
+            if (ipFilterRuleType == null) {
+                return this.rules.get(indexOf).ruleType() == IpFilterRuleType.ACCEPT;
+            } else {
+                return this.ipFilterRuleType == IpFilterRuleType.ACCEPT;
```
`this` is unnecessary.

chrisvest commented on handler/src/main/java/io/netty/handler/ipfilter/IpSubnetFilter.java:
```diff
@@ -0,0 +1,174 @@
+ * </p>
+ * <p>
+ * If you would like to explicitly take action on rejected {@link Channel}s, you should override
+ * {@link #channelRejected(ChannelHandlerContext, SocketAddress)}. </p>
```
This should say `AbstractRemoteAddressFilter#channelRejected(ChannelHandlerContext, SocketAddress)` for the javadoc reference to work.

hyperxpro replied:
Actually this is coming from https://github.com/netty/netty/blob/4be9387151d9702a17da56cffaa6e5e4c92245fd/handler/src/main/java/io/netty/handler/ipfilter/RuleBasedIpFilter.java#L32

chrisvest reviewed (commented):
I went over this again and had some more nits to pick :)

chrisvest commented on handler/src/test/java/io/netty/handler/ipfilter/IpSubnetFilterTest.java:
```diff
@@ -152,6 +152,7 @@ public void testBinarySearch() {
         ipSubnetFilterRuleList.add(buildRejectIP("200.200.200.200", 32));
         ipSubnetFilterRuleList.add(buildRejectIP("108.0.0.0", 4));
         ipSubnetFilterRuleList.add(buildRejectIP("10.10.10.10", 8));
+        ipSubnetFilterRuleList.add(buildRejectIP("2001:db8:abcd:0000::", 52));
```
Mixed version IPs should be in a separate test, as otherwise _this_ test will now fail because of the new exception thrown from the constructor, right?

hyperxpro replied:
Forgot to remove :p

chrisvest commented:
@netty-bot test this please

chrisvest reviewed (approved):
I think this looks good now.

hyperxpro commented:
Added mixed IPv4 and IPv6 and updated JavaDoc. :D

hyperxpro commented:
I am waiting for PR #10500 to be merged before I finalized now.

hyperxpro commented:
@franz1981 In meanwhile, can you tell how exactly you want benchmark code to be so you can add it to the suite?

hyperxpro commented:
```java
import io.netty.handler.ipfilter.IpFilterRuleType;
import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.BenchmarkMode;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.Setup;
import org.openjdk.jmh.annotations.State;
import org.openjdk.jmh.annotations.TearDown;
import org.openjdk.jmh.annotations.Warmup;
import org.openjdk.jmh.infra.Blackhole;
import org.openjdk.jmh.runner.Runner;
import org.openjdk.jmh.runner.RunnerException;
import org.openjdk.jmh.runner.options.Options;
import org.openjdk.jmh.runner.options.OptionsBuilder;

import java.net.InetSocketAddress;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

@State(Scope.Benchmark)
@BenchmarkMode(Mode.Throughput)
@Warmup(iterations = 5)
public class BenchmarkPerf {

    private List<IpSubnetFilterRule> ipFilterRulesListUnsorted;
    private List<IpSubnetFilterRule> ipFilterRulesListSorted;
    private InetSocketAddress remoteAddress;

    public static void main(String[] args) throws RunnerException {
        Options opt = new OptionsBuilder()
                .include(BenchmarkPerf.class.getSimpleName())
                .forks(5)
                .addProfiler("gc")
                .build();

        new Runner(opt).run();
    }

    @Setup
    public void setup() {
        ipFilterRulesListUnsorted = new ArrayList<>();
        ipFilterRulesListSorted = new ArrayList<>();
        remoteAddress = new InetSocketAddress("127.0.0.1", 9110);

        // 0.0.0.0
        // 1.1.1.1
        // 2.2.2.2
        // X.X.X.X
        for (int i = 0; i < 255; i++) {
            ipFilterRulesListUnsorted.add(new IpSubnetFilterRule(i + "." + i + "." + i + "." + i, 32, IpFilterRuleType.REJECT));
        }

        // 100.200.225.0
        // 100.200.225.1
        // 100.200.225.2
        // 100.200.225.X
        for (int i = 0; i < 255; i++) {
            ipFilterRulesListUnsorted.add(new IpSubnetFilterRule("100.200.225." + i, 32, IpFilterRuleType.REJECT));
        }

        ipFilterRulesListUnsorted.add(new IpSubnetFilterRule("127.0.0.1", 32, IpFilterRuleType.REJECT));
        ipFilterRulesListSorted.addAll(ipFilterRulesListUnsorted);
        Collections.sort(ipFilterRulesListSorted);
    }

    @Benchmark
    public void forLoop() {
        for (IpSubnetFilterRule rule : ipFilterRulesListUnsorted) {
            if (rule.matches(remoteAddress)) {
                return;
            }
        }
        throw new IllegalArgumentException("IP Address was not found");
    }

    @Benchmark
    public void binarySearchAndGetElement(Blackhole blackhole) {
        try {
            blackhole.consume(this.ipFilterRulesListSorted.get(Collections.binarySearch(this.ipFilterRulesListSorted, remoteAddress,
                    IpSubnetFilterRuleComparator.INSTANCE)).ruleType());
        } catch (ArrayIndexOutOfBoundsException ex) {
            throw new IllegalArgumentException("IP Address was not found");
        }
    }

    @Benchmark
    public void binarySearch(Blackhole blackhole) {
        try {
            blackhole.consume(Collections.binarySearch(this.ipFilterRulesListSorted, remoteAddress, IpSubnetFilterRuleComparator.INSTANCE));
        } catch (ArrayIndexOutOfBoundsException ex) {
            throw new IllegalArgumentException("IP Address was not found");
        }
    }

    @TearDown
    public void clean() {
        this.ipFilterRulesListSorted.clear();
        this.ipFilterRulesListUnsorted.clear();
    }
}

```

hyperxpro commented:
@chrisvest Having separate lists for IPv4/6, we can handle both IP versions without any issue. Also, separate lists will give advantage in binary search because IPv4 is 32-bit Integer and IPv6 is 128-bit Integer and their LOW and HIGH ranges are very big. Binary search will take a lot more time in that case.

Just added `instanceOf` checks and it's alright. The rest base algorithms are the same.

PTAL and lemme know.

chrisvest commented:
> Also, separate lists will give advantage in binary search because IPv4 is 32-bit Integer and IPv6 is 128-bit Integer and their LOW and HIGH ranges are very big. Binary search will take a lot more time in that case.

Binary search time complexity only depends on the number of elements in the list, not the domain of values being compared.

chrisvest commented:
@netty-bot test this please

hyperxpro commented:
> > Also, separate lists will give advantage in binary search because IPv4 is 32-bit Integer and IPv6 is 128-bit Integer and their LOW and HIGH ranges are very big. Binary search will take a lot more time in that case.
> 
> Binary search time complexity only depends on the number of elements in the list, not the domain of values being compared.

Same but the real reason why I added code to remove overlapping subnet was if we had 2.0.0.0/8 and few more IP addresses in this range like 2.1.1.1/32, 2.3.52.12/32.. etc. Binary search won't work. Debugged it hard and found Binary search was jumping high on ranges. So to make things simple, I added sorting for CIDR. Afterward, it worked perfectly.

hyperxpro commented:
Updated with`NetUtils#ipv4AddressToInt(Inet4Address)`.

Please run the test last time and we're good to go for merge.

normanmaurer commented:
@netty-bot test this please

hyperxpro commented:
```
13:22:54 [netty-centos6-java8-prb] $ /bin/sh -xe /tmp/jenkins6078476321631526049.sh
13:22:54 + docker-compose -f docker/docker-compose.yaml -f docker/docker-compose.centos-6.18.yaml -p netty-centos6-java8-prb run --rm test-leak
13:22:55 Creating netty-centos6-java8-prb_runtime-setup_1 ... 
13:22:55 
Creating netty-centos6-java8-prb_runtime-setup_1 ... error
13:22:55 ERROR: for netty-centos6-java8-prb_runtime-setup_1  Cannot start service runtime-setup: network netty-centos6-java8-prb_default not found
13:22:55 
13:22:55 ERROR: for runtime-setup  Cannot start service runtime-setup: network netty-centos6-java8-prb_default not found
13:22:55 Encountered errors while bringing up the project.
13:22:55 Build step 'Execute shell' marked build as failure
```

I don't remember messing with Docker. What could be the reason for failure?

normanmaurer commented:
@netty-bot test this please

hyperxpro commented:
Good to go for a merge finally. :D

normanmaurer commented:
@netty-bot test this please

normanmaurer commented:
@hyperxpro thanks a lot again..

hyperxpro commented:
Huge thanks to @franz1981 and @chrisvest for the help. :D
